[Feature/#56] 동영상 동기화 로직 구현

LURKS02 opened the pull request:
## 관련 이슈

- #56

## ✅ 완료 및 수정 내역

- SocketProvider에서 동기화 로직 분리 -> Synchronizer / Repository
- 해시 동기화 코드 구현

## 🛠️ 테스트 방법
- 시뮬레이터 2개로 테스트 했습니다. 디버깅이 잘 안되어서 나중에 메인 로직과 합쳐졌을 때 다시 테스트 해야 할 것 같습니다.<br>
- (시뮬레이터 2개가 통신 중 하나가 종료 후 다시 연결되었을 때 동영상 동기화가 되는지 테스트)

## 📝 리뷰 노트
```mermaid
sequenceDiagram
    loop [Until allSynced]
    A->>+B: 해시 값 전달: hash([String: String])
    A->>+C: 해시 값 전달: hash([String: String])
    A->>+D: 해시 값 전달: hash([String: String])
    B-->>-A: 해시 비교 정보 전달: hashMatch([fileName1: .missing])
    C-->>-A: 해시 비교 정보 전달: hashMatch([fileName2: .additional])
    D-->>-A: 해시 비교 정보 전달: hashMatch([fileName1: .missing, fileName2: .additional])
    Note right of A: sendSyncFlags: [B: false, C: false, D: false]<br/>receiveSyncFlags: [B: false, C: false, D: false]
    A->>+B: missing 파일 전달: shareResource(fileName1)
    A->>+C: additional 파일 요구: SyncMessage.request([fileName2])
    A->>+D: missing 파일 전달: shareResource(fileName1)
    A->>+D: additional 파일 요구: SyncMessage.request([fileName2])
    B-->>-A: sendResponse 전송
    Note right of A: sendSyncFlags: [B: true, C: false, D: false]<br/>receiveSyncFlags: [B: true, C: false, D: false]
    C-->>A: requested 파일 전달: shareResource(fileName2)
    C-->>-A: receiveResponse 전송
    Note right of A: sendSyncFlags: [B: true, C: true, D: false]<br/>receiveSyncFlags: [B: true, C: true, D: false]
    D-->>-A: sendResponse 전송
    Note right of A: sendSyncFlags: [B: true, C: true, D: true]<br/>receiveSyncFlags: [B: true, C: true, D: false]
    D-->>A: requested 파일 전달: shareResource(fileName2)
    D-->>-A: receiveResponse 전송
    Note right of A: sendSyncFlags: [B: true, C: true, D: true]<br/>receiveSyncFlags: [B: true, C: true, D: true]
    end
    A-->A: isSynchronized.send(Void)
    A-->B: completed
    B-->B: isSynchronized.send(Void)
    A-->C: completed
    C-->C: isSynchronized.send(Void)
    A-->D: completed
    D-->D: isSynchronized.send(Void)


    %% A -- hash([String: String]) --> C 
    %% A -- hash([String: String]) --> D
    %% B -- hash matching info (ex. [fileName1: .missing, fileName2: .additional]) --> B
```

### HashCondition
```swift
missing: 해당 파일이 없음
additional: 해당 파일이 추가적으로 존재함
```
해시 값을 비교하여 각 파일의 상태를 남깁니다.

### SyncMessage
```swift
hash([String: String]): 폴더 해시 정보를 피어들에게 전송합니다.
hashMatch([String: HashCondition]): 해시 비교 정보를 발신자에게 전송합니다.
request([String]): additional 파일들에 대해 전송을 요청합니다.
sendResponse: 전송받은 쪽에서 전송이 완료되었음을 알립니다.
receiveResponse: 받은 쪽에서 모든 수신이 완료되었음을 알립니다.
completed: 모든 동기화가 완료되고 다음 화면으로 넘어갈 수 있음을 알립니다.
```

### sendResource
기존 sendResource 로직에서 continuation leak이 발생해서 우선 주석처리 해두고 단순 전송으로 변경했습니다.<br>
단순 전송을 하더라도 편집 화면으로 넘어갈 때 동기화가 진행될 예정이므로 작동에는 문제 없지 않을까 싶습니다.... <br>
어떻게 할지 확인해주시면 반영하여 머지하도록 하겠습니다<br>
( + 그리고 receiveResource 주고받는 부분에서 뭔가 잘못 작성한 것 같은데 추후 수정하겠습니다.. 아직 테스트가 더 필요합니다 ㅠㅠ)

## 📱 스크린샷(선택)

https://github.com/user-attachments/assets/e24370a8-a7f3-42fe-a13d-156398ecff32

LURKS02 committed:
[REFACTOR]: 동영상 삭제 시점을 애플리케이션 종료가 아닌 실행으로 변경

LURKS02 committed:
[REFACTOR]: FileSystemManager에서 해시 관련 코드 제거

LURKS02 committed:
[REFACTOR]: 임시 폴더에 동일한 파일이 존재할 경우 동영상 추가가 되지 않도록 변경

LURKS02 committed:
[FEAT]: 동기화를 담당하는 Synchronizer 구현

- compareToLocal: 파라미터의 해시와 로컬 해시의 비교값 리턴
- collectHashes: 동영상 폴더의 해시 정보 수집
- calculateFileHash: 파일 해시값 계산

LURKS02 committed:
[FEAT]: 해시 비교 결과인 HashCondition enum 구현

- missing: 해당 해시 값이 없음
- additional: 해당 해시 값이 추가적으로 존재함

LURKS02 committed:
[REFACTOR]: 동기화 함수 명 변경 (synchronizeVideos -> broadcastHashes)

LURKS02 committed:
[REFACTOR]: 사용하지 않는 코드 제거

LURKS02 committed:
[REFACTOR]: ResourceSendable 이름 변경 및 shareResource 메서드 수정

- Continuation을 이용한 코드에서 leak이 난다는 경고 메시지가 있어 우선 주석 처리 해두었습니다.

LURKS02 committed:
[FEAT]: 특정 피어에게 Resource를 전송하는 shareResource 구현

LURKS02 committed:
[FEAT]: 현재 연결된 피어 ID 정보를 담는 connectedPeerIDs 구현

LURKS02 committed:
[FEAT]: sendHashes를 제거하고 범용적으로 Data를 전송하는 메서드 구현

LURKS02 committed:
[REFACTOR]: SocketProvider에서 해시 관련 코드 제거

LURKS02 committed:
[FEAT]: 해시 동작 시 주고받을 SyncMessage 구현

LURKS02 committed:
[FEAT]: 해싱 로직 구현 (Message 전송 및 수신 후 분기 처리)

LURKS02 committed:
[CHORE]: Conflict 해결

051198Hz reviewed (approved):
고생 많으셨습니다 ~~!!

051198Hz commented on Data/P2PSocket/SocketProvider.swift:
```diff
@@ -132,51 +139,81 @@ public extension SocketProvider {
+        )
+        resourceShared.send(sharedResource)
+        
+        /// 이 부분에서 continuation leak이 나는 것 같습니다.
+        /// 리소스를 공유한 뒤 최종적으로 동기화를 시키기 때문에
+        /// 현재는 주석 처리 해두었습니다.
+//        return try await withCheckedThrowingContinuation { resourceUrlContinuation in
```
P3: continuation leak은 혹시 무엇인가요..?? continuation이 resume돼지 않아 끝나지 않는 현상을 이야기 하시는건가요??

LURKS02 replied:
네 맞습니다! continuation leak이라고 경고 메시지가 뜨고 있어서 주석처리 해두었습니다

around-forest reviewed (approved):
고생 많으셨습니다!
Sequence Diagram이 있으니 이해가 쉽네요!

P3: 개인적으로 UseCase의 이름을 VideoSharingUseCase처럼 보다 구체적으로 수정해서 추후 생성될 Video Edit 관련 UseCase과 구분하기 쉽게 하는것도 좋지 않을까 싶기는 합니다!

jungseokyoung-cloud commented on Data/P2PSocket/SocketProvidable.swift:
```diff
@@ -10,9 +10,11 @@ import Foundation
+    SocketResourceSendable, DataSendable { }
 
 public protocol SocketAdvertiseable {
+    var connectedPeerIDs: [String] { get }
```
P3: 이건 `Browsable`에 connectedPeers()메서드가 있는데 프로퍼티로 하나 더 생성하신 이유는 뭔가요?

LURKS02 replied:
다른 것으로 착각했습니다 수정했습니다!

jungseokyoung-cloud commented on Data/P2PSocket/SocketProvidable.swift:
```diff
@@ -49,15 +51,21 @@ public protocol SocketDisconnectable {
 public protocol SocketResourceSendable {
 	var resourceShared: PassthroughSubject<SharedResource, Never> { get }
 	
+    func shareResource(
```
P2: 현재 받을 때와 보낼때 네이밍이 `shareResource`다 보니 좀 헷갈릴 수 있다는 우려가 생기긴 합니다. 
메서드가 아닌 주석을 보고 어떤 메서드인지 유추했기 때문에 뭔가 좀더 명확했으면 합니다.

LURKS02 replied:
리소스 전송하는 메서드 sendResource로 수정했습니다.

jungseokyoung-cloud commented on Data/P2PSocket/SocketProvider.swift:
```diff
@@ -18,7 +18,15 @@ public final class SocketProvider: NSObject, SocketProvidable {
+    
+    public var connectedPeerIDs: [String] {
+        session.connectedPeers.compactMap { peerID in
+            MCPeerIDStorage.shared.peerIDByIdentifier.first {
```
P1: 연결된 모든 peer이기 때문에 `connected`된 사람과 `pending`인 사람만 추출해야 하는거 아닌가요??

LURKS02 replied:
위에 수정하면서 반영했습니다!

jungseokyoung-cloud commented on Data/P2PSocket/SocketProvider.swift:
```diff
@@ -132,51 +139,81 @@ public extension SocketProvider {
+        url localURL: URL,
+        resourceName: String,
+        to peerID: String,
+        completion: (((any Error)?) -> Void)? = nil
```
P3: 정말 정말 사소한건데, continuation을 활용하면 async 메서드로 바꿀 수 있지 않을까요?

LURKS02 replied:
async로 바꿔서 테스트했습니다!

jungseokyoung-cloud commented on Data/P2PSocket/SocketProvider.swift:
```diff
@@ -132,51 +139,81 @@ public extension SocketProvider {
+//                    withCompletionHandler: handler
+//                )
+//                
+//                guard progress == nil else { return }
```
P2: 해당 에러는 저도 찾아보니까 resume이 한번도 호출되지 않을 때 발생한다고 하더라고요! 아마 해당 경우가 아닐까 생각합니다.

jungseokyoung-cloud commented on Core/Utilities/Synchronizer.swift:
```diff
@@ -0,0 +1,59 @@
+        let localHashes = collectHashes()
+        var result: [String: HashCondition] = [:]
+        
+        for (fileName, _) in localHashes where hashes[fileName] == nil {
```
P3: 사소한거지만 이건 filter와 같은 고차함수를 사용할 수 있지 않을까요~?

LURKS02 replied:
고차함수 활용하도록 수정했습니다.

jungseokyoung-cloud commented on Data/Data/SharingVideoRepository.swift:
```diff
@@ -28,9 +37,29 @@ public final class SharingVideoRepository: SharingVideoRepositoryInterface {
-            .store(in: &cancellables)
+        socketProvider.dataShared
+            .sink { [weak self] (data, peerID) in
+                guard let data = try? JSONDecoder().decode(SyncMessage.self, from: data) else { return }
```
P2: `guard let self else { return }`을 앞으로 분리하면 switch 문의 case마다 별도로 self바인딩을 할필요가 없을꺼 같습니다!

LURKS02 replied:
수정했습니다!

jungseokyoung-cloud commented on Data/Data/SharingVideoRepository.swift:
```diff
@@ -39,6 +68,79 @@ private extension SharingVideoRepository {
+    }
+    
+    func sendResponse(isRequested: Bool, to peerID: String) {
+        let responseFlag = isRequested ? SyncMessage.receiveResponse : SyncMessage.sendResponse
```
P2: `receiveMessage`는 언제 보내나요??

LURKS02 replied:
로직이 약간 수정되어서 receiveMessage는 requestCompletionFlag로 바뀌었습니다!

jungseokyoung-cloud commented on Data/Data/SharingVideoRepository.swift:
```diff
@@ -39,6 +68,79 @@ private extension SharingVideoRepository {
+    
+    func synchronize(with result: [String: HashCondition], to peerID: String) {
+        if result.isEmpty {
+            sendSyncFlags[peerID] = true
```
P2: 해당 메서드는 다른 유저에게 받은 것이 싱크가 맞나 확인해야 하는데, sendSyncFlags를 여기서 설정하는게 맞나요?? 
보내는 시점이 아닌 받는 시점에 해서 의문입니다.

LURKS02 replied:
다른 유저에게 받은 것이 싱크가 맞나 확인하는 작업인데, 이 작업을 리더(다음 버튼을 누른 사람)가 담당합니다. Flag를 관리하는 것 또한 리더의 역할이고 result에는 싱크가 맞지 않는 파일 정보만 들어가서, result가 없으면 싱크가 맞는 것이고 따라서 해당 peer의 Flag들을 true로 설정해주는 코드입니다.

LURKS02 replied:
로직이 조금 수정되었는데요, 최종적으로는 이렇습니다

- 다른 유저가 missing인 파일이 있다는 정보가 있을 경우
```
리더는 해당 유저에게 missing 파일을 전송합니다. (await)
이 작업이 끝나면 해당 유저의 sendSyncFlag를 true로 설정합니다.
```
- 다른 유저가 additional인 파일이 있다는 정보가 있을 경우
```
리더는 해당 유저에게 파일을 요청합니다. (request)
해당 유저가 리더에게 필요한 파일을 전송합니다.
파일 전송이 종료된 후 해당 유저가 리더에게 전송이 종료되었다는 플래그를 보냅니다. (requestCompletionFlag)
리더는 이 플래그를 받아 해당 유저의 receiveSyncFlag를 true로 설정합니다.
```

jungseokyoung-cloud reviewed (approved):
고생하셨습니다! 몇가지 궁금사항 질문 남겼습니다~!

LURKS02 committed:
[REFACTOR]: connectedPeerIDs 제거하고 connectedPeers 사용

LURKS02 committed:
[STYLE]: 리소스 전송 메서드 이름을 shareResource -> sendResource로 변경

LURKS02 committed:
[REFACTOR]: connected 와 pending 상태의 동료만 처리하도록 변경

LURKS02 committed:
[CHORE]: 빌드 오류 해결 - CRDT 의존성 제거

LURKS02 committed:
[REFACTOR]: 고차 함수 적용

LURKS02 committed:
[REFACTOR] completion 대신 continuation 사용

LURKS02 committed:
[REFACTOR]: 동기화 로직 변경

- 보내는 경우 바로 sendSyncFlag 업데이트
- 받는 경우 requestCompletionFlag가 돌아올 때 receiveSyncFlag 업데이트

LURKS02 committed:
[REFACTOR]: 이상이 없을 경우 플래그 설정하는 코드 추가
